pymc3/variational/inference.py

```diff
@@ -0,0 +1,139 @@
+import numpy as np
+import theano
+import pymc3 as pm
+from .approximations import Advi, BaseApproximation
+from .advi import adagrad_optimizer
+from tqdm import trange
+
+
+APPROXIMATIONS = {
+    'advi': Advi,
+}
+
+
+def approximate(n=10000, population=None, local_vars=None,
+                optimizer=None, method='advi', samples=1, pi=1,
+                callbacks=None, learning_rate=.001, epsilon=.1,
+                model=None, more_params=None, more_updates=None,
+                *args, **kwargs):
+    """Interface for efficient variational inference with gradient
+    variance reduction described in [4]_
+
+
+    PyMC3 supports the following variational inference methods:
+        1) Automatic differentiation variational inference (ADVI) [2]_
+
+    Parameters
+    ----------
+    n : int
+        number of training iterations
+    population : dict
+        maps observed_RV to its population size
+        if not provided defaults to full population
+        Note: population size is not `shape[0]` but `size` attr
+    local_vars : dict[Variable->(mu, rho)]
+        maps random variable to mu and rho
+        for posterior parametrization it is used for Autoencoding Variational Bayes
+        (AEVB; Kingma and Welling, 2014)[1]_
+    optimizer : callable
+        optional custom optimizer to be called in the following way:
+            :code:`updates = optimizer(-elbo, list_of_params)`
+    method : str|Approximation
+        string description of approximation to be used or
+        Approximation instance used to calculate elbo and provide for shared params
+    samples : int|Tensor
+        number of Monte Carlo samples used for approximation,
+        defaults to 1
+    pi : float|Tensor
+        pi in [0;1] reweighting constant for KL divergence
+        this trick was described in [3]_ for fine-tuning minibatch ADVI
+    callbacks : list[callable]
+        callables that will be called in the following way
+            :code:`callback(Approximation, elbo_history, i)`
+    learning_rate : float
+        learning rate for adagrad optimizer,
+        it will be ignored if optimizer is passed
+    epsilon : float - epsilon for adagrad optimizer,
+        it will be ignored if optimizer is passed
+    model : Model
+        Probabilistic model if function is called out of context
+    more_params : list
+        optional more parameters for computing gradients
+    more_updates : dict
+        more updates are included in step function
+    args : additional args that will be passed to Approximation
+    kwargs : additional kwargs that will be passed to Approximation
+
+    Returns
+    -------
+    (Approximation, elbo_history)
```

Comment:
I think SVGD can compute ELBO because we have samples from variational posterior. Is it wrong?